### PR TITLE
Bug 986510 - [Messages app] Replace menu icon with settings icon when no messages in app

### PR DESCRIPTION
**Pull request by yvtheja**

### CODE REVIEW

**Comment by try-server-hook** [comment]
[ythej Vishnu Teja (ythej)](https://github.com/ythej) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=e5ecdc2dad89)

File: apps/sms/js/thread_list_ui.js [hunk 1]
```diff
@@ -498,7 +505,9 @@ var ThreadListUI = {
     var removeWhenEmpty = empty ? 'remove' : 'add';
 
     ThreadListUI.noMessages.classList[removeWhenEmpty]('hide');
+    ThreadListUI.settingsButton.classList[removeWhenEmpty]('hide');
     ThreadListUI.container.classList[addWhenEmpty]('hide');
+    ThreadListUI.optionsButton.classList[addWhenEmpty]('hide');
```
Comment (julienw) [inline apps/sms/js/thread_list_ui.js]: how about toggling a new class "is-empty" on the "thread-list" container and changing the CSS accordingly?

Note that "classList.toggle" can take a second argument that is boolean, nowadays :)
```javascript
// use !! to force a boolean, because if it's undefined it will be as if it's absent.
panel.classList.toggle('threadlist-is-empty', !!empty);
```

@azasypkin maybe you'll have a better idea than me for the class name? I feel like `is-empty` is too small and thus too generic. But to be BEM-like we should add a "ThreadList" class and a "isEmpty" class to have a selector ".ThreadList.isEmpty". What do you think?

Comment (azasypkin) [inline apps/sms/js/thread_list_ui.js]: > @azasypkin maybe you'll have a better idea than me for the class name? I feel like is-empty is too small and thus too generic.

Yeah, ```is-empty``` is a way too generic, totally agree.

> But to be BEM-like we should add a "ThreadList" class and a "isEmpty" class to have a selector ".ThreadList.isEmpty". What do you think?

We already have ```threadlist``` prefix for other ThreadListUI-related selectors (eg. [1]) so I think it's good to have one. So either ```threadlist-is-empty``` or ```threadlist-no-messages``` sounds good. I'm also fine if we eventually start using BEM-like state modifiers - ```threadlist--empty``` as we discussed in [2]. 

[1] https://github.com/mozilla-b2g/gaia/blob/master/apps/sms/style/sms.css#L144
[2] https://etherpad.mozilla.org/sms-code-style-guidelines

Comment (ythej) [inline apps/sms/js/thread_list_ui.js]: @azasypkin @julienw how about 'threadlist-no-threads' ?

Comment (julienw) [inline apps/sms/js/thread_list_ui.js]: I like "threadlist-is-empty" because it represents a state with the "is" word. It's a convention. But I don't have a strong feeling.

**Comment by try-server-hook** [comment]
[ythej Vishnu Teja (ythej)](https://github.com/ythej) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=5b77ec37975f)

File: apps/sms/test/unit/thread_list_ui_test.js [hunk 2]
```diff
@@ -173,16 +163,6 @@ suite('thread_list_ui', function() {
...
-    test('show settings/cancel options when list is empty', function() {
-      ThreadListUI.setEmpty(true);
-      ThreadListUI.showOptions();
-
-      var optionItems = MockOptionMenu.calls[0].items;
-      assert.equal(optionItems.length, 2);
-      assert.equal(optionItems[0].l10nId, 'settings');
-      assert.equal(optionItems[1].l10nId, 'cancel');
-    });
-
```
Comment (ythej) [inline apps/sms/test/unit/thread_list_ui_test.js]: Hi @julienw ,
I tried to add a test case like 'test(settings button is displayed when list is empty...' and with assert like 'assert.equal(settingsButton.style.display, 'flex'); and assert.equal(optionsButton.style.display, 'none');' .But there was an error as 'expected ' ' to be 'flex/none' ' . Is this test case needed?
Thank you.

**Comment by try-server-hook** [comment]
[ythej Vishnu Teja (ythej)](https://github.com/ythej) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=b15e8599dc74)

File: apps/sms/js/thread_list_ui.js [hunk 3]
```diff
@@ -515,7 +521,7 @@ var ThreadListUI = {
     };
 
     // Add delete option when list is not empty
-    if (ThreadListUI.noMessages.classList.contains('hide')) {
+    if (!ThreadListUI.container.classList.contains('hide')) {
```
Comment (julienw) [inline apps/sms/js/thread_list_ui.js]: do we still need this condition ? I think showOptions is not actionnable when the container is empty, with your patch?

File: apps/sms/style/sms.css [hunk 4]
```diff
@@ -1062,3 +1062,19 @@ html[dir="rtl"] .contact-header aside.pack-end {
...
+#threads-settings-button, .threadlist-is-empty #threads-options-button {
+  display: none !important;
+}
+
+.threadlist-is-empty #threads-settings-button {
+  display: flex !important;
+}
+
+.threadlist-is-empty #threads-no-messages {
+  display: flex;
```
Comment (julienw) [inline apps/sms/style/sms.css]: nit: `display: unset`

File: apps/sms/style/sms.css [hunk 5]
```diff
@@ -1062,3 +1062,19 @@ html[dir="rtl"] .contact-header aside.pack-end {
...
+#threads-no-messages, .threadlist-is-empty #threads-container {
+  display: none;
+}
+
+#threads-settings-button, .threadlist-is-empty #threads-options-button {
+  display: none !important;
+}
+
+.threadlist-is-empty #threads-settings-button {
+  display: flex !important;
```
Comment (julienw) [inline apps/sms/style/sms.css]: nit: add a comment explaining we use "flex" to center the content

File: apps/sms/style/sms.css [hunk 6]
```diff
@@ -1062,3 +1062,19 @@ html[dir="rtl"] .contact-header aside.pack-end {
 .editable-select-mode [contenteditable] {
   -moz-user-select: text;
 }
+
+#threads-no-messages, .threadlist-is-empty #threads-container {
+  display: none;
+}
+
+#threads-settings-button, .threadlist-is-empty #threads-options-button {
+  display: none !important;
```
Comment (julienw) [inline apps/sms/style/sms.css]: nit: add a comment explaining we use "!important" to override scoped style added by the gaia-header component

File: apps/sms/style/sms.css [hunk 7]
```diff
@@ -1062,3 +1062,19 @@ html[dir="rtl"] .contact-header aside.pack-end {
 .editable-select-mode [contenteditable] {
   -moz-user-select: text;
 }
+
```
Comment (julienw) [inline apps/sms/style/sms.css]: nit: add a comment to separate this css block; like `/* empty threadlist layout */`

File: apps/sms/test/unit/thread_list_ui_test.js (hunk 2 above)
Comment (julienw) [inline apps/sms/test/unit/thread_list_ui_test.js]: no you can't really do that because we don't apply the CSS files in unit tests.

You can try doing a integration tests, but it's not really easy so I won't force you to do one.

File: build/csslint/xfail.list [hunk 8]
```diff
@@ -51,7 +51,7 @@ apps/settings/style/settings_large.css 1 1
 apps/settings/style/screen_lock.css 0 2
 apps/settings/style/downloads.css 0 3
 apps/sms/style/root.css 0 1
-apps/sms/style/sms.css 0 4
+apps/sms/style/sms.css 0 6
```
Comment (julienw) [inline build/csslint/xfail.list]: you need to rebase as this conflicts with another patch

File: apps/sms/style/sms.css (hunk 4 above)
Comment (ythej) [inline apps/sms/style/sms.css]: @julienw , with 'display: unset', the icon is not correctly placed.But we are able to set it to right position by giving it some margin.But we need to set !important for the margin also.I saw that we used 'flex' in shared/gaiaheader so I implemented it .Do I need to do with 'display: unset' ?

File: apps/sms/test/unit/thread_list_ui_test.js (hunk 2 above)
Comment (ythej) [inline apps/sms/test/unit/thread_list_ui_test.js]: Thank you :)

File: apps/sms/style/sms.css (hunk 4 above)
Comment (julienw) [inline apps/sms/style/sms.css]: ok, we really want "display: block", sorry

Comment (ythej) [inline apps/sms/style/sms.css]: @julienw , okay, I will try with it :)

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

File: apps/sms/js/thread_list_ui.js [hunk 9]
```diff
@@ -514,13 +520,10 @@ var ThreadListUI = {
...
-    if (ThreadListUI.noMessages.classList.contains('hide')) {
-      params.items.unshift({
-        l10nId: 'selectThreads-label',
-        method: this.startEdit.bind(this)
-      });
-    }
+    params.items.unshift({
+      l10nId: 'selectThreads-label',
+      method: this.startEdit.bind(this)
+    });
```
Comment (julienw) [inline apps/sms/js/thread_list_ui.js]: and because we don't have the if condition anymore, you can directly add it in the previous object params, you don't need to use params.items.unshift.

**Comment by try-server-hook** [comment]
[ythej Vishnu Teja (ythej)](https://github.com/ythej) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=b506bc9bf958)

File: apps/sms/js/thread_list_ui.js [hunk 10]
```diff
@@ -508,20 +514,15 @@ var ThreadListUI = {
         method: function oSettings() {
           ActivityPicker.openSettings();
         }
+      },{
+        l10nId: 'selectThreads-label',
+        method: this.startEdit.bind(this)
```
Comment (julienw) [inline apps/sms/js/thread_list_ui.js]: sorry, it should be at the first place in the array :/

**Comment by julienw** [comment]
looks like the gaia unit tests are failing: https://treeherder.mozilla.org/#/jobs?repo=gaia-try&revision=b506bc9bf958

**Comment by try-server-hook** [comment]
[ythej Vishnu Teja (ythej)](https://github.com/ythej) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=44a9d97254c6)

**Comment by try-server-hook** [comment]
[julienw Julien Wajsberg (julienw)](https://github.com/julienw) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=e7304aa38be0)

**Comment by julienw** [comment]
I missed this earlier, but you need to remove  the file "npm-debug.log" from your patch.

**Comment by try-server-hook** [comment]
[ythej Vishnu Teja (ythej)](https://github.com/ythej) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=428a012216e7)
